### PR TITLE
Add safe pandas eval tool

### PR DESCRIPTION
**Pull request by nfcampos**
- [ ] whitelist FROM clause to only allow selecting from the names of the data frames passed in or subqueries (which are themselves subject to same whitelisting)

Screenshot 2023-10-03 at 13 31 57

### CODE REVIEW

File: libs/langchain/langchain/agents/agent_toolkits/pandas/base.py [hunk 1]
```diff
@@ -22,16 +22,19 @@
 from langchain.schema.language_model import BaseLanguageModel
 from langchain.schema.messages import SystemMessage
 from langchain.tools import BaseTool
+from langchain.tools.pandas_eval.pandas_eval import PandasEvalTool
 from langchain.tools.python.tool import PythonAstREPLTool
 
 
 def _get_multi_prompt(
     dfs: List[Any],
+    llm: BaseLanguageModel,
```
Comment (baskaryan) [inline libs/langchain/langchain/agents/agent_toolkits/pandas/base.py:31]: should make this optional

**Review by baskaryan — COMMENTED** [review]
v cool

File: libs/langchain/langchain/agents/agent_toolkits/pandas/base.py (hunk 1 above)
Comment (nfcampos) [reply libs/langchain/langchain/agents/agent_toolkits/pandas/base.py:31]: Why?

**Comment by eyurtsev** [comment]
reviewing

**Comment by eyurtsev** [comment]
@nfcampos I'm on board with using duckdb for this. 

The latency doubles with the extra call, and it's pretty significant now. 

What do you think if we tried to re-write this agent more generically to use duckdb?

We also discussed changing the configuration to disable file system reads, and we can add some additional sanitization to create a whitelist that restricts which sql keywords can be used.

**Pull request closed without merging**